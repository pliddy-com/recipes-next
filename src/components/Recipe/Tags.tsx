import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';

import TagButtons from 'components/TagButtons';

import { RecipeTagsCollection, Tag } from 'schema';

type TagsProps = {
  collection?: RecipeTagsCollection;
};

const Tags = ({ collection }: TagsProps) => {
  if (!collection) return null;

  const { items } = collection ?? {};
  const tags = items as Array<Tag>;

  return (
    <Box>
      <Typography variant="h2" gutterBottom>
        Tags
      </Typography>
      <TagButtons tags={tags} />
    </Box>
  );
};

export default Tags;
