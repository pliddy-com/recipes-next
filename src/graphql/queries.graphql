# Query for Homepage RecipeCollection (all recipes)

query homePageQuery {
  recipeCollection(order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe collection based on filter

query recipeCollection($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC) {
    total
    items {
      ...RecipeSummary
    }
  }
}

# Query for Recipe page

query recipePage($where: RecipeFilter) {
  recipeCollection(where: $where, order: title_ASC, limit: 1) {
    total
    items {
      ...RecipeDefault
    }
  }
}

# Query for List Page RecipeCollection (all recipes matching tag)

query listPageQuery($where: TagFilter) {
  tagCollection(where: $where, limit: 1) {
    total
    items {
      slug
      title
      linkedFrom {
        recipeCollection(limit: 24) {
          total
          items {
            ...RecipeSummary
          }
        }
      }
    }
  }
}

# Query for recipe slugs

query recipeSlugsCollection($where: RecipeFilter) {
  recipeCollection(where: $where) {
    items {
      slug
    }
  }
}

# query for category slugs

query categorySlugsCollection($where: TagFilter) {
  tagCollection(where: $where) {
    items {
      slug
    }
  }
}

# query for category slugs

query tagSlugsCollection($where: TagFilter) {
  tagCollection(where: $where) {
    items {
      slug
      linkedFrom {
        recipeCollection {
          total
        }
      }
    }
  }
}

# query for Taxonomy with Category & Subcategory levels

query taxonomyCollection($where: TaxonomyFilter) {
  taxonomyCollection(where: $where, limit: 1) {
    items {
      ...TaxonomyDefault

      tag {
        ...TagDefault
      }

      childrenCollection(limit: 24) {
        total
        items {
          ... on Tag {
            ...TagDefault

            linkedFrom {
              recipeCollection(limit: 1) {
                total
              }
            }
          }
          ...TaxonomyWithChildren
        }
      }
    }
  }
}
