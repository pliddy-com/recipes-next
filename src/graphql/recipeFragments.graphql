# Fragments with default fields for Recipe-specific sections:
#   * Ingredients
#   * Instructions
#   * Image
#   * Recipe

fragment IngredientsDefault on IngredientSection {
  sys {
    id
  }
  title
  slug
  label
  ingredientList
}

fragment InstructionsDefault on InstructionSection {
  sys {
    id
  }
  title
  slug
  label
  instructionList
}

fragment ImageDefault on Asset {
  sys {
    id
  }
  __typename
  title
  description
  contentType
  fileName
  size
  url
  height
  width
}

fragment DescriptionDefault on Recipe {
  description {
    json
    links {
      entries {
        inline {
          __typename
          sys {
            id
            __typename
          }
        }
        hyperlink {
          __typename
          sys {
            id
            __typename
          }
        }
        block {
          __typename

          sys {
            id
            __typename
          }
        }
      }
      assets {
        hyperlink {
          __typename
          title
        }
        block {
          __typename
          title
        }
      }
    }
  }
}

fragment RecipeDefault on Recipe {
  sys {
    id
  }
  __typename
  title
  slug
  description {
    json
  }
  abstract
  image {
    ...ImageDefault
  }
  ingredientsCollection(limit: 10) {
    items {
      ...IngredientsDefault
    }
  }
  equipment
  instructionsCollection(limit: 10) {
    items {
      ...InstructionsDefault
    }
  }
  notes
  tagsCollection(limit: 10) {
    items {
      ...TagDefault
    }
  }
}
