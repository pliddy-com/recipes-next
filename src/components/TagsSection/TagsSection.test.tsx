// import testing-library methods
import { render } from '@testing-library/react';

// add custom jest matchers from jest-dom
import '@testing-library/jest-dom';

// import the component to test
import Tags from 'components/TagsSection/TagsSection';
import { TagDefaultFragment } from 'types/generated/graphql';

describe('Tags', () => {
  describe('when there is a properly formatted tags property', () => {
    const tags = [
      {
        sys: {
          id: 'sysid-1',
          __typename: 'Sys',
        },
        __typename: 'Tag',
        title: 'Tag 1',
        slug: 'tag-1',
      },
      {
        sys: {
          id: 'sysid-2',
          __typename: 'Sys',
        },
        __typename: 'Tag',
        title: 'Tag 2',
        slug: 'tag-2',
      },
    ];

    it('it renders the notes section', () => {
      const testId = 'tags-section';
      const tagButtonsId = 'tag-buttons';
      const expectedTitle = tags[0]?.title;
      const expectedSlug = tags[0]?.slug;

      const { queryByText, queryByTestId } = render(
        <Tags tags={tags as (TagDefaultFragment | null)[]} />
      );

      // assert that the component is rendered
      const component = queryByTestId(testId);

      // assert that the tag title is rendered
      expect(queryByText('Tags')).toBeInTheDocument();

      // assert that a TagButtons component is rendered
      // const tagButtons = queryByTestId(tagButtonsId);
      expect(queryByTestId(tagButtonsId)).toBeInTheDocument();

      // assert that tag href links are correct
      const tag = queryByText(expectedTitle)?.closest('a');
      expect(tag).toHaveAttribute('href', `/tag/${expectedSlug}`);

      // assert that the component matches the existing snapshot
      expect(component).toMatchSnapshot();
    });
  });

  describe('when there is not a properly formatted tags property', () => {
    it('it does not render the tags section if there is no content', () => {
      const tags: (TagDefaultFragment | null)[] = [];
      const expectedLabel = 'Tags';

      const { queryByText } = render(<Tags tags={tags} />);

      const title = queryByText(expectedLabel);

      expect(title).toBeNull();
    });
  });
});
