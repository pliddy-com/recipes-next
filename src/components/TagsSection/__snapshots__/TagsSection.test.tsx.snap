// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags when there is a properly formatted tags property it renders the notes section 1`] = `
<div
  class="MuiBox-root css-0"
  data-testid="tags-section"
>
  <h2
    class="MuiTypography-root MuiTypography-h2 MuiTypography-gutterBottom css-i1hrn3-MuiTypography-root"
  >
    Tags
  </h2>
  <div
    class="css-180l46g-MuiStack-root"
    data-testid="tag-buttons"
  >
    <a
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-wmndcn-MuiButtonBase-root-MuiButton-root"
      href="/tag/tag-1"
      role="button"
      tabindex="0"
    >
       Tag 1 
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </a>
    <a
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-wmndcn-MuiButtonBase-root-MuiButton-root"
      href="/tag/tag-2"
      role="button"
      tabindex="0"
    >
       Tag 2 
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </a>
  </div>
</div>
`;
